hack/generate-yamls.sh

```diff
@@ -58,7 +58,11 @@ readonly MONITORING_TRACE_JAEGER_IN_MEM_YAML=${YAML_OUTPUT_DIR}/monitoring-traci
 readonly MONITORING_LOG_ELASTICSEARCH_YAML=${YAML_OUTPUT_DIR}/monitoring-logs-elasticsearch.yaml
 
 # Flags for all ko commands
-readonly KO_YAML_FLAGS="-P ${KO_FLAGS}"
+if [[ ${KO_DOCKER_REPO} == *"gcr.io"* ]]; then
+  readonly KO_YAML_FLAGS="-P ${KO_FLAGS}"
+else
+  readonly KO_YAML_FLAGS="-B ${KO_FLAGS}"
```

Comment:
This is probably fine, but I do want to point out that given that our image names are fairly generic "controller", "autoscaler", "webhook", etc. there is a high chance that these names may clash with other Knative projects. Without a path prefix it may be difficult to images apart if you push to the wrong KO_DOCKER_REPO.

Reply:
It's safer to match the host part of KO_DOCKER_REPO.

```
KO_YAML_FLAGS="-P"
[[ "${KO_DOCKER_REPO}" == gcr.io/* ]] && KO_YAML_FLAGS="-B"
readonly KO_YAML_FLAGS="${KO_YAML_FLAGS} ${KO_FLAGS}"
```

Reply:
+1, this is a good point.

Reply:
Don't we also have an option to append the SHA to the actual name? I think that'd solve that issue.

Reply:
it's the md5, and this is the default behavior (no `-P`, no `-B`)

Reply:
It works fine for both gcr and other registries without -P or -B.
For example, this is the image name generated index.docker.io/someone's name/autoscaler-b41733a3e5694e8fc64777bdfd8e791a@sha256:f9fb8cb806450fd430b1ca855e8e5c11c984429c9f3df8df77983a77ee98aee1.
The easy fix will be change readonly KO_YAML_FLAGS="-P ${KO_FLAGS}" into readonly KO_YAML_FLAGS="${KO_FLAGS}", or replace all the KO_YAML_FLAGS with KO_FLAGS.

Reply:
This is a breaking change to the way images are published.

Reply:
I guess your focus is just the E2E tests, and this script is also used for releases (which I doubt you're running since the entire release process is tied to gcr). I would prefer not changing the settings for gcr, and using the default settings (no -P, no -B) for non-gcr registries. Since our releases are published to gcr, it won't affect them, but would provide the desired functionality for E2E tests.

Reply:
I see.
-P for release with gcr,
no arg for others.